academic_observatory_workflows/workflows/tests/test_openalex_telescope.py

```diff
@@ -0,0 +1,560 @@
+                "transform": ["upload_transformed"],
+                "upload_transformed": ["bq_load_partition"],
+                "bq_load_partition": ["bq_delete_old"],
+                "bq_delete_old": ["bq_append_new"],
+                "bq_append_new": ["cleanup"],
+                "cleanup": [],
+            },
+            dag,
+        )
+
+    def test_dag_load(self):
+        """Test that the OpenAlex DAG can be loaded from a DAG bag.
+        :return: None
+        """
+
+        with ObservatoryEnvironment().create():
+            dag_file = os.path.join(module_file_path("academic_observatory_workflows.dags"), "openalex_telescope.py")
+            self.assert_dag_load("openalex", dag_file)
+
+    @patch("academic_observatory_workflows.workflows.openalex_telescope.aws_to_google_cloud_storage_transfer")
+    @patch("academic_observatory_workflows.workflows.openalex_telescope.boto3.client")
+    def test_telescope(self, mock_client, mock_transfer):
+        """Test the OpenAlex telescope end to end.
+        :return: None.
+        """
+        # Setup Observatory environment
+        env = ObservatoryEnvironment(self.project_id, self.data_location)
+        dataset_id = env.add_dataset()
+
+        # Setup Telescope
+        telescope = OpenAlexTelescope(dataset_id=dataset_id)
+        dag = telescope.make_dag()
+
+        # Create the Observatory environment and run tests
+        with env.create():
+            # Add connection
+            conn = Connection(
+                conn_id=OpenAlexTelescope.AIRFLOW_CONN_AWS, uri="aws://UWLA41aAhdja:AJLD91saAJSKAL0AjAhkaka@"
+            )
+            env.add_connection(conn)
+
+            release = None  # prevent linting error when using release variable in second run
+            for run in [self.first_run, self.second_run]:
+                with self.subTest(run=run):
+                    with env.create_dag_run(dag, run["execution_date"]) as dag_run:
+                        # Test that all dependencies are specified: no error should be thrown
+                        env.run_task(telescope.check_dependencies.__name__)
+                        start_date, end_date, first_release = telescope.get_release_info(
+                            next_execution_date=pendulum.today("UTC"),
+                            dag=dag,
+                            dag_run=dag_run,
+                        )
+                        if run == self.first_run:
+                            self.assertEqual(dag.default_args["start_date"], start_date)
+                            self.assertEqual(pendulum.today("UTC") - timedelta(days=1), end_date)
+                            self.assertTrue(first_release)
+                        else:
+                            self.assertEqual(release.end_date + timedelta(days=1), start_date)
+                            self.assertEqual(pendulum.today("UTC") - timedelta(days=1), end_date)
+                            self.assertFalse(first_release)
+
```

Comment:
I personally think that utilising loops and if statements for large blocks of unit tests makes them difficult to follow, and that it is preferable for the tests to be more verbose (with the tradeoff of duplication of unit testing code), so that it is easier and quicker to understand how a particular part of the system is tested. The unit tests are partly a recipe for another programmer to learn how the system works.

That is just my opinion though, so don't feel like you have to change it.

Reply:
Mmm yeah, thank you for the feedback. I think you're right and will reformat a bit.